cmd/flux-mini: add mini run/submit

garlick opened the pull request:
As discussed in #2379, here's a first cut on a `flux mini` command (with `run` and `submit` subcommands), for early review.

It borrows big hunks shamelessly from @SteVwonder's `flux-jobspec.py`.

SteVwonder commented on src/cmd/flux-mini.py:
```diff
@@ -0,0 +1,335 @@
+        help="Number of cpus to allocate per task",
+    )
+    parser.add_argument(
+        "-t", "--time", help="Time limit: min[:sec] or [days-]hours:min:sec"
```
Since `mini run` doesn't need to abide by Slurm's time format, do we want to be consistent with `flux run` and use the Flux Standard Duration?  [(Discussion on time in `flux run` PR)](https://github.com/flux-framework/flux-core/issues/2213)

garlick replied:
Good idea!  Just pushed a fixup.

I grabbed @trws's `parse_fsd()` from #2323 and changed it to raise a ValueError if input is not valid.  This is a good candidate function for adding to the python bindings IMHO.

trws replied:
One of the two would definitely be good.  The bindings may actually have access to the parse_fsd function in libutil already, but this way we avoid the string conversion etc.  Either way, but we should have it in flux.util or some such for sure.

SteVwonder replied:
Just added a new issue to track the `parse_fsd` function in python: https://github.com/flux-framework/flux-core/issues/2393

SteVwonder commented on src/cmd/flux-mini.py:
```diff
@@ -0,0 +1,335 @@
+        type=kv_list_split,
+        default=[{}],
+        help="List or specify environment variables to export",
+    )
```
If you drop the `append` action and set the `default={}`, then you can remove the `exports = {...}` processing in `get_environment`.  The `args.export` dictionary will already be in the format you need.

All of that assumes `--export` will only be provided on the command line one time.  If that's a bad asummption, that ignore everything I said and what you have is :ok_hand: :+1:.

garlick replied:
I just copied what was in `flux jobspec srun` here.  It does seem like supporting multiple occurrences of `--setenv`  is a plus.  However, the option was a clone of srun's and we're not tied to that here, so maybe there is room for improvement?

grondo replied:
I would avoid adding options to `flux mini run` unless they are critical at this point. I'm not sure that `--setenv` is used anywhere at this time.

dongahn commented:
This wouldn't have implication to the schedulers, would it?

grondo commented:
> Do we need some way to run more tasks than cores like slurms --oversubscribe?

Maybe eventually but not now?

In general, maybe we shouldn't take direct inspiration from Slurm commands. For instance, maybe we don't need `--oversubscribe` when we support `tasks_per_slot`?

garlick commented:
Just squashed and force- pushed an update that
* uses Flux standard duration for --time
* drops --setenv
* replaces --show-jobid, --show-events --show-exec with a --verbose option that can be specified multiple times (easier to maintain interface stability this way)

garlick commented:
I was wondering if it might improve clarity a bit if `--ntasks` must be specified if greater than 1, rather than allowing `--nnodes` to set the task count?  E.g. require
```
flux mini run --nnodes=2 --ntasks=2 cmd ...
```
rather than just
```
flux mini run --nnodes=2 cmd ...
```
Let that fail with
```
flux-mini: node count must not exceed task count
```

trws commented:
It would certainly make it more explicit, I wouldn't mind that change personally.

garlick commented:
Just rebased and squashed some fixups with minor cleanup.

I was trying to add the ability to set shell options but didn't quite get that done today.

garlick commented:
OK, latest fixup is a refactoring into a JobSpec class, SubmitCmd class, and RunCmd class.  Sorry, that change touches everything, but I think it's a good cleanup.  I hope the JobSpec class is starting to look more like something that could be factored out to a reusable module.

I did change the behavior of `--nodes` and `--ntasks` as described above (tentatively, easy to change back)

I'm pretty tentative about this change so would appreciate any feedback from people with more python experience.  I can easily drop it if it's not desirable.

Otherwise I'll continue and try to add the shell options today.

grondo commented:
> I did change the behavior of --nodes and --ntasks as described above (tentatively, easy to change back)

I agree that seems to be a good change. 

It is probably just me, but the usage and help output from python argparse based stuff is just ugly:

```
usage: flux-mini.py [-h] {run,submit} ...

positional arguments:
  {run,submit}
    run         run subcommand
    submit      submit subcommand

optional arguments:
  -h, --help    show this help message and exit
```

I guess this is probably for a future PR, but could we make the python commands behave and output help like our other commands (which are modeled after `git`, the gold standard)

It would also be nice to add a bit more info in the default help output than just `"run command"` and `"submit command"`, and also `flux mini help` doesn't work.

Finally, before we tag we may want to add a short manpage for `flux-mini(1)` since then it will appear in the listing of available commands in `flux help`.

grondo commented on src/cmd/flux-mini.py:
```diff
@@ -0,0 +1,305 @@
+        os.execvp("flux-job", attach_args)
+
+
+logging.basicConfig(level=logging.INFO, format="flux-mini: %(levelname)s %(message)s")
```
Maybe add a colon after `%(levelname)s` so messages appear as `flux-mini: ERROR: message`

SteVwonder replied:
This should already be handled by the `CLIMain` decorator [[1]](https://github.com/flux-framework/flux-core/blob/master/src/bindings/python/flux/util.py#L76).  I think the following line just needs to be tweaked to use `flux-mini` rather than `__name__` based on this open issue (https://github.com/flux-framework/flux-core/issues/2234).

grondo commented on src/cmd/flux-mini.py:
```diff
@@ -0,0 +1,305 @@
+
+@util.CLIMain(logger)
+def main():
+    parser = argparse.ArgumentParser()
```
maybe set `prog="flux mini"` for argparse so that help doesn't reference `flux-mini.py` which breaks the 4th wall of the `flux(1)` command driver ;-)

garlick replied:
Ok, applying the fix to the util class discussed in #2234, and setting the name to "flux-mini".

grondo replied:
Was this pushed, still seeing:
```console
$ flux mini 
usage: flux-mini.py [-h] {run,submit} ...
flux-mini.py: error: too few arguments
```

This is minor enough that it could wait for a future pr.

garlick replied:
Sorry, my my comment was about the logging prefix.  I had not set the `ArgumentParser(prog="flux-mini")`.  I can also override the usage string to print something more like our standard boiler plate for commands with subcommands here.

Fixup coming.

grondo replied:
Ah, sorry, my misinterpretation. I see now that it is obvious what you meant.

grondo commented on src/cmd/flux-mini.py:
```diff
@@ -0,0 +1,305 @@
+            "--cpus-per-task",
+            type=int,
+            default=1,
+            help="Number of cpus to allocate per task",
```
Totally optional: Use `metavar='N'` in these `add_argument` calls to get more traditional help output instead of e.g. `-c CPUS_PER_TASK, --cpus-per-task CPUS_PER_TASK`

grondo commented on src/cmd/flux-mini.py:
```diff
@@ -0,0 +1,305 @@
+        """
+        if type(opts) is not dict:
+            raise ValueError("shell options must be a dictionary")
+        self.jobspec["attributes"]["system"]["shell"]["options"] = opts
```
This function isn't used yet, but my feeling is that we want to be able to specify `-o` multiple times on the command line with something like `-o mpi=spectrum -o cpu-affinity=off` (or equivalently a comma separated list). Does this function completely reassign all options? We might want to add a way to incrementally add them instead.

garlick replied:
Thanks!  Good feedback as I start working on that.

SteVwonder commented on scripts/check-format:
```diff
@@ -17,9 +17,9 @@ if command -v black 2>&1 > /dev/null ; then
+            | xargs black --check --diff
     else
-        filter_pyfiles "$@" | xargs black --check
+        filter_pyfiles "$@" | xargs black --check --diff
```
Very nice addition! 👍

SteVwonder commented on src/cmd/flux-mini.py:
```diff
@@ -0,0 +1,305 @@
+        Constructor builds the minimum legal v1 jobspec.
+        Use setters to assign additional properties.
+        """
+        if type(command) is not list:
```
I think this is close to what you want, but not exactly.  Python's duck typing means you'll trim out a bunch of things that act like lists but aren't explicitly lists (e.g., `tuple`, `deque`).   ([relevant stackoverflow answer](https://stackoverflow.com/a/37842328)).   In `_create_resource`, there is an example of checking for that it is `collectionsAbc.Sequence` and not a `str`.  (Although now that I look at that again, it only works for python 3 and not python 2....grr....maybe we should wrap that functionality up into a utility function.....feel free to not include that in this PR, we can always open an issue and loop back around to it).

SteVwonder replied:
One other thing, you'll want to switch to using `isintance` rather than `type` to account for inheritance ([relevant stackoverflow post](https://stackoverflow.com/a/1549854)).

garlick replied:
OK, I'll use isinstance and try to be a bit more inclusive but hold off on the collectionsAbc stuff.

SteVwonder replied:
@trws : any insight on the "right way" to test for a thing that maps to a `list` in YAML?

I think the "Sequence not str" check above also allows buffers and other unintended types in.

trws replied:
I see this is now assertions etc.  While that’s fine, we could be using typing comments and mypy if we are doing type-checking.

As to actually telling if it will become a list, the only way I know to be sure is to actually turn the thing into yaml and back and check, since types can register their own serializers.  Just checking that it’s a sequence and not a string seems reasonable.

SteVwonder commented on src/cmd/flux-mini.py:
```diff
@@ -0,0 +1,305 @@
+        - a string in Flux Standard Duration
+        A duration of zero is interpreted as "not set".
+        """
+        if type(duration) is str:
```
python2 and python3 compatible string handling is unfortunately a bit of a bear.  To make this compatible with python2, you'll need to check `if isinstance(duration, six.string_types)` [[1]](https://six.readthedocs.io/#six.string_types).

trws replied:
Since python 2 EOL is coming imminently, it might be worth checking with our customers to see how long we need to hang on to python 2 support.  Alternately, we may require python3 and optionally additionally build python 2 bindings so we don’t have to deal with this stuff in our own commands.

SteVwonder replied:
That's a great point @trws:  just opened an issue to track the python2 EOL actions.

garlick commented:
Pushed some fixups on the review comments so far (minus prettifying help output which I struggled with a bit, and man page which I agree needs to be added).

SteVwonder commented on src/cmd/flux-mini.py:
```diff
@@ -0,0 +1,305 @@
+        """
+        Set working directory of job.
+        """
+        if type(cwd) is not str:
```
see `six.string_types` comment above

SteVwonder commented on src/cmd/flux-mini.py:
```diff
@@ -0,0 +1,305 @@
+        """
+        Set (entire) environment of job.
+        """
+        if type(environ) is not dict:
```
Would be more general to check if it is an instance of a `collections.abc.Mapping` [[1]](https://docs.python.org/2/library/collections.html#collections.Mapping)

garlick commented:
Pushed a fixup with the more relaxed type checking.

garlick commented:
Added couple more changes
* address stephen's feedback
* fix for logging prefix
* improve help descriptions of subcommands when one isn't specified
* add --setopt KEY=VAL option for setting shell options (where val can be JSON or a string)
* add --dryrun to print jobspec and exit

Still need to add option for setting output modes.

grondo commented:
> add --setopt KEY=VAL option for setting shell options (where val can be JSON or a string)

Nicely done! I think options will be common enough that `-o` should also be allowed to set options as `--setopt option=value --setopt option2=value2` starts to get a bit cumbersome.

Very minor nit on `--dryrun`: I think `--dry-run` seems to be more standard practice, (e.g. `git` commands like e.g.`git-rm(1)`)

BTW, really good idea to add `--dry-run` option. This is all turning out nice IMO.

grondo commented:
Restarting a builder that failed with the well known valgrind test failure.

garlick commented:
Squashed the fixups and rebased on current master.

Still need -O,--output and -E,--error options, manpage, and sharness test.

garlick commented:
Just pushed a man page and fixups for
* `--setattr=KEY=VAL` option to set any jobspec attr, as suggested by @grondo 
* fix to `-o, --setopt=KEY=VAL` to allow periods in keys (as path separators)
* add `--output` and `--error` options, as discussed in #2241 

On to add some tests and deprecate `flux-srun`.

SteVwonder commented on doc/man1/flux-mini.adoc:
```diff
@@ -0,0 +1,147 @@
+*-N, --nodes=N*::
+Set the number of nodes to assign to the job.  Tasks will be distributed
+evenly across the allocated nodes.  It is an error to request more nodes
+than there are tasks.  If unspecified, the minimum number of nodes
```
Maybe I'm misinterpreting, but I'm interpreting this statement to mean that the front-end tool will request the minimal number of nodes required to satisfy the user's overall request.  IIUC, if `nodes` is unspecified, the front-end tool omits `nodes` from the jobspec, and the scheduler will make it's decision without constraining the number of nodes in any way (you could get N nodes where N is the number of tasks, or 1 node, just depends on the scheduler).

garlick replied:
Right, that statement was not only misinformed, but is an inappropriate level of detail for a behavior that is configurable elsewhere.

SteVwonder commented on doc/man1/flux-mini.adoc:
```diff
@@ -0,0 +1,147 @@
+VAL may be valid JSON (bare values, objects, or arrays), otherwise VAL
+is interpreted as a string.
+
+*--dry-run*::
```
With this option now supported, is the plan to deprecate/drop the `flux jobspec` command? (Fine by me, just wondering)

garlick replied:
Since `flux mini` has deviated from the original slurm compatibility (FSD for time limit, `--nodes` doesn't alter `--ntasks`)  `flux jobspec srun` may still be useful?  There is the potential for providing jobspec generation from other convenient idioms in the future.  My vote would be to keep it and possibly refactor to use `JobSpec()` class if appropriate down the road.

SteVwonder commented on src/cmd/flux-mini.py:
```diff
@@ -218,6 +230,29 @@ def create_parser(self):
+            metavar="FILENAME",
+        )
+        parser.add_argument(
+            "--label",
```
Nit: I kinda prefer `--label-io` only because `--label` seems ambiguous as to what is being labelled.  Is the job getting a label?   Not a hill I will die on though, so please feel free to disregard.

grondo replied:
I vote --label-io as well

grondo replied:
I vote --label-io as well

garlick replied:
On the subject of option names, would it be more descriptive to rename
* `--cpus-per-task` to `--cores-per-task`?
* `--time` to `--time-limit` or `--max-duration`?

grondo replied:
I think both of those are a good idea. (for `--time`, `--time-limit` is preferable just because the terminology will be more familiar to users)

trws commented:
On Sep 25, 2019, at 1:29 PM, Mark Grondona <notifications@github.com> wrote:

It is probably just me, but the usage and help output from python argparse based stuff is just ugly:

usage: flux-mini.py [-h] {run,submit} ...

positional arguments:
  {run,submit}
    run         run subcommand
    submit      submit subcommand

optional arguments:
  -h, --help    show this help message and exit


I guess this is probably for a future PR, but could we make the python commands behave and output help like our other commands (which are modeled after git, the gold standard)

The printing is completely configurable, or replaceable, if you’re sufficiently determined.  What do you find particularly ugly about it?

grondo commented:
> The printing is completely configurable, or replaceable, if you’re sufficiently determined.  What do you find particularly ugly about it?

Yeah, I think @garlick has fixed most of my minor complaints already. Just as examples, some things that cluttered the help output included:

 * By default, argparse picks its "metavar" as underscored and capitalized version of the long argument, *and* applies it to both the short and long option. e.g.
```
  -c CORES_PER_TASK, ---cores-per-task CORES_PER_TASK
```
vs
```
  -c, --cores-per-task=N
```

 * By default sub-commands are listed as "positional arguments" in curly braces which clutters output and most users probably don't even know what that means, e.g.
```
usage: flux-mini [-h] {run, submit} ...
positional arguments:
 {run,submit}
  run     run a parallel job
  submit  submit a parallel job
```
vs the more preferable
```
usage: flux-mini [-h] COMMAND ...

Supported commands:
   run      run a parallel job
   submit  submit a parallel job
```

As you said, it seems most of this is configurable, but personally I just find the defaults a bit off putting. (I agree just personal preference though)

trws commented:
Yeah, I think @garlick<https://github.com/garlick> has fixed most of my minor complaints already. Just as examples, some things that cluttered the help output included:

  *   By default, argparse picks its "metavar" as underscored and capitalized version of the long argument, and applies it to both the short and long option. e.g.

  -c CORES_PER_TASK, ---cores-per-task CORES_PER_TASK


vs

  -c, --cores-per-task=N


Agreed, look at: https://docs.python.org/3.4/library/argparse.html#argparse.MetavarTypeHelpFormatter

Specifying that formatter will switch it from using that to giving the expected type, which it not identical but is much more my preference at least.  Would that work for you?


  *   By default sub-commands are listed as "positional arguments" in curly braces which clutters output and most users probably don't even know what that means, e.g.

usage: flux-mini [-h] {run, submit} ...
positional arguments:
 {run,submit}
  run     run a parallel job
  submit  submit a parallel job


vs the more preferable

usage: flux-mini [-h] COMMAND ...

Supported commands:
   run      run a parallel job
   submit  submit a parallel job


Have a look at: https://docs.python.org/3.4/library/argparse.html#argparse.ArgumentParser.add_subparsers

Particularly the “title” and “description” keyword arguments.  Adding ‘title=“Supported commands:”, description=“”’ should give almost exactly that output.  FWIW I prefer having the title set so they get grouped separately as sub commands rather than positional arguments as well.

garlick commented:
Thanks @trws, trying out
```diff
diff --git a/src/cmd/flux-mini.py b/src/cmd/flux-mini.py
index 84999057d..ad0003054 100755
--- a/src/cmd/flux-mini.py
+++ b/src/cmd/flux-mini.py
@@ -368,7 +368,7 @@ logger = logging.getLogger("flux-mini")
 @util.CLIMain(logger)
 def main():
     parser = argparse.ArgumentParser(prog="flux-mini")
-    subparsers = parser.add_subparsers()
+    subparsers = parser.add_subparsers(title="supported subcommands", description="")
 
     # run
     run = RunCmd()
```
yields
```
$ flux mini -h
usage: flux-mini [-h] {run,submit} ...

optional arguments:
  -h, --help    show this help message and exit

supported subcommands:

  {run,submit}
    run         run a job interactively
    submit      enqueue a job
```
an improvement I think.

garlick commented:
I made the suggested option name changes, rebased on master, and squashed the fixups.
Then I tacked on a little debugging for mysterious travis failures... (Thanks @SteVwonder for the help)

grondo commented:
Close to ready?

trws commented:
Looking good @garlick. I forgot the brace-list option thing, if you set... I think it's meta_var or metavar or something, on the subcommands to something and all the same it should replace that brace-list with that, so metavar="" or something might squash that off too if we want rid of it.

garlick commented:
Oh sheesh, having a bit of a time with travis.  I'll check in later to see what's gone wrong this time.

grondo commented:
Looks like more 3.6 issues.. error in travis logs is:
```
flux-mini: ERROR: '>' not supported between instances of 'NoneType' and 'int'
```

The TEST_INSTALL builder had a different error:
```
flux-mini: ERROR: [Errno 2] No such file or directory
```

garlick commented:
Pushed one fix: it turns out `print(flush=True)` is only a python3 thing, so replaced that with `os.stderr.flush()` which should be portable.

grondo commented:
That's annoying that python chooses to make `stderr` buffered. [posix](https://pubs.opengroup.org/onlinepubs/9699919799/functions/stdin.html) says:

> When opened, the standard error stream is not fully buffered; the standard input and standard output streams are fully buffered if and only if the stream can be determined not to refer to an interactive device.

Oh well.

grondo commented on doc/man1/flux-mini.adoc:
```diff
@@ -0,0 +1,148 @@
+-----------
+flux-mini(1) submits jobs to run under Flux.  The job consists of
+'N' copies of COMMAND launched together as a parallel job.
+If '--nprocs' is unspecified, a value of 'N=1' is assumed.
```
`--nprocs` is used here and in SYNOPSIS above instead of `ntasks`.

garlick commented:
> I was playing around on my Ubuntu 18.04 system and hit the following error:

Oh duh, in my haste to fix that python3ism, I tacked on a flush of os.stderr, not sys.stderr.  Sorry about that and thanks for poking at it!

garlick commented:
Setting merge-when-passing since @SteVwonder was going to do it yesterday afternoon but I wanted to squash the fixups first.  Now they're squashed and all comments addressed I think.
